### ISSUE TITLE
Open in the Origin path when using SaveFile

### ISSUE THREAD

**Issue by xhorak** [flatpak/xdg-desktop-portal#982]
Problem:
In Firefox where the user can choose location where to save the downloaded file (ie. `Always ask you where to save files` checked) the file is downloaded (for example to `/run/user/1000/doc/abcdef/downloadedFile.ext`) and the path `/run/user/1000/doc/abcdef` is stored by the Firefox on per site basis to show the same location again when downloading another file.

In sandboxed app the `/run/user/1000/doc/abcdef` path is shown on subsequent download requests. Since only the previously downloaded file is in the documents portal ( `/run/user/1000/doc/abcdef/downloadedFile.ext`) the subsequent download fails silently - the application does not have access to `/run/user/1000/doc/abcdef`.

Solution:
* When FileChooser SaveFile is called with `/run/user/…` path, try to lookup the path in the documents portal, when found set the path current dirrectory to its Origin path. 
* When not found (most likely because the previous file is in the documents portal but not the folder) try to find the corresponding Origin path by removing the filename from the `/run/...` path, when a match is found use the Origin path minus filename path for the file chooser as current directory
* When it fails, use the provided path (if path does not start with `/run/user/`) or home directory.

**Comment by matthiasclasen** [comment]
Lets clarify what problem we are trying to solve here, exactly.

Is the issue that firefox does not know about the host location (and thus shows the document store path /run/usr/...) ?

Or is the issue that the portal save dialog does not open the host location ?

**Comment by matthiasclasen** [comment]
To solve the first one, we would have to add more information to the response that the portal sends back (add a 'host-path' field to the response vardict).

To solve the second, we'd have to do something like what you describe.

**Comment by xhorak** [comment]
> Is the issue that firefox does not know about the host location (and thus shows the document store path /run/usr/...) ?
Yes and even worse. It has remembered just the directory where the previous file was saved to. And this directory is not in the documents portal - so the download fails when user just hits enter.

We could save the host path if the SaveFile portal provided it. This is related to `content-prefs` and `browser.download.lastDir` preference.

If you would pass the host path from the firefox to the SaveFile portal and it would change the directory in the dialog to the host path we should be fine - because the new documents portal entry will be made on the return of the file chooser dialog, right?

> Or is the issue that the portal save dialog does not open the host location ?
Currently the dialog opens in the `/run/user/1000/doc/abcd` which is not meant to be for writing.

It all depends if flatpak portal is willing to pass the host-path to the flatpak app. I also must admit that the getting host-path from the portal would require some none trivial code changes on the firefox site since some of the code is written in the js where we have to detect the flatpak environment, etc. 

Also I can imagine that there are some other applications which also would like to store the previous location too - like editors or so. If all that would be implemented in the portal it would make the app developers life easier.

**Comment by matthiasclasen** [comment]
I would be willing to consider passing the host pass from the portal to the app, for display purposes.

Passing it the other way seems much more iffy

**Comment by xhorak** [comment]
> I would be willing to consider passing the host pass from the portal to the app, for display purposes.
> 
> Passing it the other way seems much more iffy

I'm afraid that won't help much. Maybe only for the OpenFolder which is used for custom download dir which is showed up as:
![image](https://user-images.githubusercontent.com/2932697/231453963-d1e1933a-bbe6-43c6-87c5-122078a72a68.png)
but that's different issue.

But from the portal documentation I see that you can set the `current_folder` to suggest folder to save to:
https://docs.flatpak.org/en/latest/portal-api-reference.html#gdbus-method-org-freedesktop-portal-FileChooser.SaveFile

**Comment by matthiasclasen** [comment]
Ah yeah, maybe that works. Can `current_folder` just be any host path? I guess its up to the portal whether it wants to allow that, or ignore it.

**Comment by matthiasclasen** [comment]
https://github.com/flatpak/xdg-desktop-portal/pull/1000 has some work to improve things

**Comment by Erick555** [comment]
Similar issues were reported earlier:
https://github.com/flatpak/xdg-desktop-portal/issues/886
https://github.com/flatpak/xdg-desktop-portal/issues/973

**Comment by ssokolow** [comment]
As I've argued in the other issues, the problem is that the path handling in the file chooser portal is asymmetrical.

The file chooser portal will happily convert host paths given by the user into documents portal paths, but will not then reverse the process and convert document portal paths given by the application into host paths for display.

In one instance, I likened it to a browser's address bar performing a DNS lookup on `google.com` as expected, but then forcing the displayed URL to use `https://142.251.33.174/` instead of `https://google.com/`. Neglecting to reverse documents portal paths to host paths when displaying to the user is equally un-intuitive and, as with IP addresses in the era of the TLS SNI and the `Host:` header, breaks things if the user tries to modify the path/URL.

More practically, it makes things painful for the user in any application that does its own more fine-grained path-remembering, such as a browser remembering last-used paths of a per domain basis, an image editor like GIMP remembering last-used paths on a per-tab basis (Remember that GIMP remembers Save and Export paths independently), or an IDE remembering last-used paths on a per-project basis.
